### PR TITLE
Fix AQI index calculator

### PR DESCRIPTION
**Pull request by freekode**
# What does this implement/fix? 

Without conversion important fractions were lost.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/hm3301/caqi_calculator.h [hunk 1]
```diff
@@ -37,7 +37,7 @@ class CAQICalculator : public AbstractAQICalculator {
     int conc_lo = array[grid_index][0];
     int conc_hi = array[grid_index][1];
 
-    int aqi = ((aqi_hi - aqi_lo) / (conc_hi - conc_lo)) * (value - conc_lo) + aqi_lo;
+    int aqi = ((aqi_hi - aqi_lo) / (double) (conc_hi - conc_lo)) * (value - conc_lo) + aqi_lo;
```
Comment (oxan) [inline esphome/components/hm3301/caqi_calculator.h]: We use single-precision floats for performance reasons:
```suggestion
    int aqi = ((aqi_hi - aqi_lo) / (float) (conc_hi - conc_lo)) * (value - conc_lo) + aqi_lo;
```

**Review by oxan — CHANGES_REQUESTED** [review]

File: esphome/components/hm3301/caqi_calculator.h (hunk 1 above)
Comment (oxan) [reply esphome/components/hm3301/caqi_calculator.h]: Actually, given that the end result is converted to an integer anyway, rearranging the order should also fix the calculation without incurring the performance cost of floats:
```suggestion
    int aqi = (value - conc_lo) * (aqi_hi - aqi_lo) / (conc_hi - conc_lo) + aqi_lo;
```

Comment (freekode) [reply esphome/components/hm3301/caqi_calculator.h]: Thanks, fixed

**Review by oxan — APPROVED** [review]

**Comment by oxan** [comment]
Thank you for your contribution!

**Comment by martgras** [comment]
This is not directly related to this PR  but I  build a version without the arduino dependency.
I don't want to submit a new PR because I don't have the device and cannot test the code 

Do you mind testing

```yaml

external_components:
  - source: github://martgras/esphome@hm3301-idf
    components: ["hm3301"]

substitutions:
  devicename: hm-test
esphome:
  name: ${devicename}

esp32:
  board: esp32dev
  framework:
    type: esp-idf
logger:
  level: VERY_VERBOSE

i2c:
  sda: 21
  scl: 22

sensor:
  - platform: hm3301
    address: 0x40
    pm_1_0:
      name: "PM1.0"
    pm_2_5:
      name: "PM2.5"
    pm_10_0:
      name: "PM10.0"
    aqi:
      name: "AQI"
      calculation_type: "CAQI"
```

The code is at https://github.com/martgras/esphome/tree/hm3301-idf/esphome/components/hm3301
If it works feel free to use it for follow-up PR

**Comment by freekode** [comment]
@martgras I don't have ESP32, I tested it on my ESP8266.
Looks like it's working well, here my [commit](https://github.com/freekode/esp-at-home/commit/01272f6fabf38ef48cadd94f939e64cf3b57b2f5) with updated config. So you can check did I test it correctly.

**Comment by martgras** [comment]
Thanks for testing - the new PR is https://github.com/esphome/esphome/pull/2745

**Pull request merged**